x-pack/plugins/rule_registry/server/utils/create_lifecycle_executor.ts

```diff
@@ -0,0 +1,326 @@
+
+import { Logger } from '@kbn/logging';
+import { getOrElse } from 'fp-ts/lib/Either';
+import * as rt from 'io-ts';
+import { Mutable } from 'utility-types';
+import { v4 } from 'uuid';
+import {
+  AlertExecutorOptions,
+  AlertInstance,
+  AlertInstanceContext,
+  AlertInstanceState,
+  AlertTypeParams,
+  AlertTypeState,
+} from '../../../alerting/server';
+import { ParsedTechnicalFields, parseTechnicalFields } from '../../common/parse_technical_fields';
+import {
+  ALERT_DURATION,
+  ALERT_END,
+  ALERT_ID,
+  ALERT_START,
+  ALERT_STATUS,
+  ALERT_UUID,
+  EVENT_ACTION,
+  EVENT_KIND,
+  RULE_UUID,
+  TIMESTAMP,
+} from '../../common/technical_rule_data_field_names';
+import { RuleDataClient } from '../rule_data_client';
+import { AlertExecutorOptionsWithExtraServices } from '../types';
+import { getRuleData } from './get_rule_executor_data';
+
+type LifecycleAlertService<
+  InstanceState extends AlertInstanceState = never,
+  InstanceContext extends AlertInstanceContext = never,
+  ActionGroupIds extends string = never
+> = (alert: {
+  id: string;
+  fields: Record<string, unknown>;
+}) => AlertInstance<InstanceState, InstanceContext, ActionGroupIds>;
+
+export interface LifecycleAlertServices<
+  InstanceState extends AlertInstanceState = never,
+  InstanceContext extends AlertInstanceContext = never,
+  ActionGroupIds extends string = never
+> {
+  alertWithLifecycle: LifecycleAlertService<InstanceState, InstanceContext, ActionGroupIds>;
+}
+
+export type LifecycleRuleExecutor<
+  Params extends AlertTypeParams = never,
+  State extends AlertTypeState = never,
+  InstanceState extends AlertInstanceState = never,
+  InstanceContext extends AlertInstanceContext = never,
+  ActionGroupIds extends string = never
+> = (
+  options: AlertExecutorOptionsWithExtraServices<
+    Params,
+    State,
+    InstanceState,
+    InstanceContext,
+    ActionGroupIds,
+    LifecycleAlertServices<InstanceState, InstanceContext, ActionGroupIds>
+  >
+) => Promise<State | void>;
+
+const trackedAlertStateRt = rt.type({
+  alertId: rt.string,
+  alertUuid: rt.string,
+  started: rt.string,
+});
+
+export type TrackedLifecycleAlertState = rt.TypeOf<typeof trackedAlertStateRt>;
+
+const alertTypeStateRt = <State extends AlertTypeState>() =>
+  rt.record(rt.string, rt.unknown) as rt.Type<State, State, unknown>;
+
+const wrappedStateRt = <State extends AlertTypeState>() =>
+  rt.type({
+    wrapped: alertTypeStateRt<State>(),
+    trackedAlerts: rt.record(rt.string, trackedAlertStateRt),
+  });
+
+/**
+ * This is redefined instead of derived from above `wrappedStateRt` because
+ * there's no easy way to instantiate generic values such as the runtime type
+ * factory function.
+ */
+export type WrappedLifecycleRuleState<State extends AlertTypeState> = AlertTypeState & {
+  wrapped: State | void;
+  trackedAlerts: Record<string, TrackedLifecycleAlertState>;
+};
+
+export const createLifecycleExecutor = (logger: Logger, ruleDataClient: RuleDataClient) => <
```

Comment:
This is just the lighter-weight extracted executor factory previously located in the `createLifecycleRuleTypeFactory` helper.